File: core/src/adictive/games/components/FloorEffectComponent.java
package adictive.games.components;

import com.badlogic.ashley.core.Component;
import com.badlogic.ashley.core.Engine;
import com.badlogic.ashley.core.Entity;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.TextureRegion;

public class FloorEffectComponent implements Component {
    public static final TextureRegion iceRegion = new TextureRegion(new Texture(Gdx.files.internal("data/iceFloor.png")));
    public static final TextureRegion dirtRegion = new TextureRegion(new Texture(Gdx.files.internal("data/dirtFloor.png")));
    public static final TextureRegion accelRegion = new TextureRegion(new Texture(Gdx.files.internal("data/floorAccel.png")));
    public static final float WIDTH = 1f;
    public static final float HEIGHT = 1f;

    public static final byte ICE = 1;
    public static final byte DIRT = 2;
    public static final byte ACCELERATOR = 3;

    public static Entity addNew(Engine engine, float x, float y, byte type) {
        Entity block = create((int) x, (int) y, type);
        engine.addEntity(block);
        return block;
    }

    public static Entity create(int x, int y, byte type) {
        Entity block = new Entity();
        block.flags = type;

        TextureComponent textureComponent = new TextureComponent();

        switch (type) {
            case ICE:
                textureComponent.region = iceRegion;
                break;
            case DIRT:
                textureComponent.region = dirtRegion;
                break;
            case ACCELERATOR:
                textureComponent.region = accelRegion;
                break;
        }

        block.add(textureComponent);

        TransformComponent transformComponent = new TransformComponent();
        transformComponent.size.set(WIDTH, HEIGHT);
        transformComponent.pos.set(x, y, 0f);
        block.add(transformComponent);

        BoundsComponent boundsComponent = new BoundsComponent();
        boundsComponent.bounds.set(0, 0, WIDTH, HEIGHT);
        block.add(boundsComponent);

        block.add(new FloorEffectComponent());
        return block;
    }
}

File: core/src/adictive/games/components/FloorEffectComponent.kt
package adictive.games.components

import com.badlogic.ashley.core.Component
import com.badlogic.ashley.core.Engine
import com.badlogic.ashley.core.Entity
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.TextureRegion

class FloorEffectComponent : Component {
    companion object {
        val iceRegion = TextureRegion(Texture(Gdx.files.internal("data/iceFloor.png")))
        val dirtRegion = TextureRegion(Texture(Gdx.files.internal("data/dirtFloor.png")))
        val accelRegion = TextureRegion(Texture(Gdx.files.internal("data/floorAccel.png")))
        val WIDTH = 1f
        val HEIGHT = 1f

        val ICE: Byte = 1
        val DIRT: Byte = 2
        val ACCELERATOR: Byte = 3

        fun addNew(engine: Engine, x: Float, y: Float, type: Byte): Entity {
            val block = create(x.toInt(), y.toInt(), type)
            engine.addEntity(block)
            return block
        }

        fun create(x: Int, y: Int, type: Byte): Entity {
            val block = Entity()
            block.flags = type.toInt()

            val textureComponent = TextureComponent()

            when (type) {
                ICE -> textureComponent.region = iceRegion
                DIRT -> textureComponent.region = dirtRegion
                ACCELERATOR -> textureComponent.region = accelRegion
            }

            block.add(textureComponent)

            val transformComponent = TransformComponent()
            transformComponent.size.set(WIDTH, HEIGHT)
            transformComponent.pos.set(x.toFloat(), y.toFloat(), 0f)
            block.add(transformComponent)

            val boundsComponent = BoundsComponent()
            boundsComponent.bounds.set(0f, 0f, WIDTH, HEIGHT)
            block.add(boundsComponent)

            block.add(FloorEffectComponent())
            return block
        }
    }
}
